resources/patches/ruby-stomp/connection_loss_reconnect_fix.patch

```diff
@@ -0,0 +1,125 @@
+From ed5da4d1d857c427bf995e50c9baa25befaa3175 Mon Sep 17 00:00:00 2001
+From: Reid Vandewiele <reid@puppetlabs.com>
+Date: Thu, 1 Jun 2017 16:22:08 -0700
+Subject: [PATCH] Ensure IO#gets exits in a reasonable time
+
+According to the notes on heartbeats, `max_hbrlck_fails` should be
+enabled in order to determine when heartbeats stop arriving. This is
+required because the receive thread of the MCO daemon spends most of its
+time waiting for something to come in over the wire. That ties up the
+"read lock" used to provide exclusive access to the underlying network
+socket.
+
+However, the blocking receive should consume heartbeats and update the
+"last received" timestamp (`@lr`). The heartbeat read thread also
+consumes pings coming in over the wire in case the process is not
+blocking on receive.
+
+Normal operation means that a missed hbrlck is normal, but two in a row
+should be abnormal (the heartbeat thread may try to read and get locked,
+but the receive thread should read the heartbeat before the heartbeat
+thread resumes - networking timing could impact this in bad ways - and
+reset the "last received" timestamp which will trigger resetting the
+`read_lock_count` and `lock_fail_count` the next time the heartbeat
+thread wakes).
+
+When the `lock_fail_count` exceeds `@max_hbrlck_fails`, the heartbeat
+thread closes the socket and kills the heartbeat threads. The
+expectation is that this should cause the `IO#gets` call in the receive
+call to return. That doesn't happen until TCP timeouts occur, which can
+take minutes or hours. When the TCP timeout occurs, it appears to leave
+the main thread in an unexpected state that sometimes results in SIGABRT
+and is unreachable via the broker.
+
+To exit `gets` early, trigger an exception in the `receive` thread.
+That interrupts the `gets` call and is caught in the `transmit` or
+`__old_receive` method and initiates a reconnect.
+---
+ lib/connection/heartbeats.rb |  6 +++---
+ lib/connection/netio.rb      | 19 +++++++++++++++----
+ lib/stomp/connection.rb      |  1 +
+ 3 files changed, 19 insertions(+), 7 deletions(-)
+
+diff --git a/lib/connection/heartbeats.rb b/lib/connection/heartbeats.rb
+index ae350dd..dc0cb15 100644
+--- a/lib/connection/heartbeats.rb
++++ b/lib/connection/heartbeats.rb
+@@ -232,9 +232,9 @@ module Stomp
+             # Retry on max lock fails.  Different logic in order to avoid a deadlock.
+             if (@max_hbrlck_fails > 0 && lock_fail_count >= @max_hbrlck_fails)
+               # This is an attempt at a connection retry.
+-              begin
+-                @socket.close # Attempt a forced close
+-              rescue
++              @gets_semaphore.synchronize do
++                @getst.raise(Errno::EBADF.new) if @getst rescue nil # kill the socket reading thread if exists
```

Comment:
Generally yes. Maybe you'd like to spend some time in https://github.com/stompgem/stomp.

Less snarkily, `.nil?` seems overly defensive since there's only 2 places it's set, but I don't see a reason against it. Also, the next step after this is to exit the thread, so the blanket rescue is to ensure cleanup happens correctly.